File: code/com/leetcode/solutions/NO_0018_四数之和/Solution.js
/**
 * 参照三数之和，多加一次嵌套
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function (nums, target) {
    nums.sort((a, b) => { return a - b });
    let ans = [];
    for (let l = 0; l < nums.length - 3; l++) {
        if (nums[l] === nums[l - 1]) continue;
        for (let i = l + 1; i < nums.length - 2; i++) {
            if (i > l + 1 && nums[i] === nums[i - 1]) continue;
            let j = i + 1;
            let k = nums.length - 1;
            while (j < k) {
                let target_temp = nums[l] + nums[i] + nums[j] + nums[k];
                if (target_temp == target) {
                    ans.push([nums[l], nums[i], nums[j], nums[k]]);
                    while (nums[--k] === nums[k + 1]);
                    while (nums[++j] === nums[j - 1]);
                }
                else if (target_temp > target) {
                    while (nums[--k] === nums[k + 1]);
                }
                else {
                    while (nums[++j] === nums[j - 1]);
                }
            }
        }
    }
    return ans;
};

File: code/com/leetcode/solutions/NO_0018_四数之和/Solution.cs
using System;

public class Solution
{

    /*
    * 
    * 思路：将数组元素4个一组进行组合，然后求和，留下与目标值相等的组合
    * 其他：
   */

    public IList<IList<int>> FourSum(int[] nums, int target)
    {

        IList<IList<int>> res = new List<IList<int>>();
        if (nums.Length <= 4)
        {
            int sum = 0;
            IList<int> rr = new List<int>();
            foreach (int v in nums)
            {
                sum += v;
                rr.Add(v);
            }
            if (sum == target)
            {
                res.Add(rr);
            }

        }
        else
        {
            //组合
            List<int[]> ListCombination = GetCombination(nums, 4);


            foreach (int[] v in ListCombination)
            {
                int sum = 0;
                IList<int> rr = new List<int>();
                foreach (int vv in nums)
                {
                    sum += vv;
                    rr.Add(vv);
                }
                if (sum == target)
                {
                    res.Add(rr);
                }
            }
        }
        return res;
    }

    private static void GetCombination(ref List<T[]> list, T[] t, int n, int m, int[] b, int M)
    {
        for (int i = n; i >= m; i--)
        {
            b[m - 1] = i - 1;
            if (m > 1)
            {
                GetCombination(ref list, t, i - 1, m - 1, b, M);
            }
            else
            {
                if (list == null)
                {
                    list = new List<T[]>();
                }
                T[] temp = new T[M];
                for (int j = 0; j < b.Length; j++)
                {
                    temp[j] = t[b[j]];
                }
                list.Add(temp);
            }
        }
    }

    public static List<T[]> GetCombination(T[] t, int n)
    {
        if (t.Length < n)
        {
            return null;
        }
        int[] temp = new int[n];
        List<T[]> list = new List<T[]>();
        GetCombination(ref list, t, t.Length, n, temp, n);
        return list;
    }


}
